### DISCUSSION TITLE
Allow project.yaml to be centralized (be an Asset).

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#8642] [Ideas]
In it's current implementation, project config and it's related config (`useProjectConfigFile`, `allowAdminChanges`), don't really fit our workflow.

Here is our normal workflow for a site in development:

- Developers work on the codebase locally, deploying to the dev site
- Content admins work in the CP on the dev site, and do much of the content modeling (therefore `allowAdminChanges` must be `true`).
- Developers will frequently refresh their local databases from the remote dev env

Having `'useProjectConfigFile' => true` for the remove env doesn't really make sense, as it is essentially writing into a black hole, as that is a sourced file that will get overwritten on any deploy.

So the best we're left with is `'useProjectConfigFile' => true` for our local environment, and every time we pull a new database we're prompted for changes, where we would always "use current project config". This will update our sourced project.yaml, which is nice as we now have a sourced representation of our schema, but isn't particularly useful.

If we were able to define our `project.yaml` as an asset (similar to https://github.com/craftcms/cms/issues/2584), I could put it on an S3 volume, where there could now be a 2-way sync of admin changes.

**Comment by brandonkelly** [comment]
> Developers will frequently refresh their local databases from the remote dev env

This is where your current workflow is breaking down. If you want `allowAdminChanges` to be `true` on your remote environment, then you should be treating it as another development environment. Install Git on it, clone the same Git repo you are sharing across other environments, and any time you make config changes there, make sure you’re committing and pushing those `project.yaml` changes up.

You can still pull DB backups down from there, but only for the purpose of getting the latest _content_ changes. Any time you pull a DB backup without grabbing the project.yaml changes along with it, you are running the risk of having a DB that is out of sync with the project config, which will lead to lots of issues down the road.

**Comment by timkelty** [comment]
> Install Git on it, clone the same Git repo you are sharing across other environments, and any time you make config changes there, make sure you’re committing those project.yaml changes up.

Makes sense in theory, but doesn't really fit into our deployment processes, or any I've heard of (most deployment workflows only ever have the remote _pull_ from the git origin).

It would still require a developer to commit those changes to the remote repo, which kind of defeats the purpose.

I think the best we could do is have a writable `project.yaml` as an "shared" file (in Capistrano parlance), that developers can manually sync to/from their local envs. This is really the same concept as storing it in s3, but a bit more manual/explicit.

**Comment by timkelty** [comment]
> Any time you pull a DB backup without grabbing the project.yaml changes along with it, you are running the risk of having a DB that is out of sync with the project config, which will lead to lots of issues down the road.

Currently to avoid this, if we pull a new db, we _always_ choose "use project config". Which is essentially the same as not using project.yaml, but I like having the changes tracked in the repo.

**Comment by brandonkelly** [comment]
Yeah I’m not saying this is a bad idea or that we won’t consider it. Just pointing out that as of the way things work _now_, you’re doing it wrong :)

**Comment by timkelty** [comment]
Totally agree.

And to be clear - we _only_ have `useProjectConfigFile' => true` on for our local env currently (and are always using db changes when we pull the db), so we're basically just using it to source admin changes in our repo via `project.yaml` – though it may become useful for spinning up new environments, though those usually need the content as well, which would mean a full dump.
